compiler/rustc_lexer/src/lib.rs

```diff
@@ -179,24 +181,27 @@ pub enum DocStyle {
 /// `rustc_ast::ast::LitKind`).
 #[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
 pub enum LiteralKind {
-    /// "12_u8", "0o100", "0b120i99", "1f32".
+    /// `12_u8`, `0o100`, `0b120i99`, `1f32`.
     Int { base: Base, empty_int: bool },
-    /// "12.34f32", "1e3", but not "1f32".
+    /// `12.34f32`, `1e3`, but not `1f32`.
     Float { base: Base, empty_exponent: bool },
-    /// "'a'", "'\\'", "'''", "';"
+    /// `'a'`, `'\\'`, `'''`, `';`
     Char { terminated: bool },
-    /// "b'a'", "b'\\'", "b'''", "b';"
+    /// `b'a'`, `b'\\'`, `b'''`, `b';`
     Byte { terminated: bool },
-    /// ""abc"", ""abc"
+    /// `"abc"`, `"abc`
     Str { terminated: bool },
-    /// "b"abc"", "b"abc"
+    /// `b"abc"`, `b"abc`
     ByteStr { terminated: bool },
     /// `c"abc"`, `c"abc`
     CStr { terminated: bool },
-    /// "r"abc"", "r#"abc"#", "r####"ab"###"c"####", "r#"a". `None` indicates
+    /// `#"abc"#`, `#"a`, `##"a"#`. `None` indicates no closing quote.
+    /// Allows fewer hashes to close the string to support older editions.
+    GuardedStr { n_start_hashes: Option<NonZeroU8>, n_end_hashes: u8 },
```

Comment:
This confuses me.
- "`None` indicates no closing quote." Wouldn't it make sense for `n_end_hashes` to have the `Option` instead of `n_start_hashes`?
- "Allows fewer hashes to close the string to support older editions."  What does this mean? This is a new feature, how are older editions relevant? Why is an unequal number of hashes allowed in older editions?

Reply:
> Why is an unequal number of hashes allowed in older editions?

For backwards compatibility. Consider code in the wild with `#"foo"`. Since lexing is mostly edition independent, we have to store some token that can be broken into `#`,`"foo"` for edition 2021 or `#"foo"<invalid>` on edition 2024.

> Wouldn't it make sense for `n_end_hashes` to have the `Option` instead of `n_start_hashes`?

That would make more sense. However, moving the `Option` causes the enum to grow in size, from two bytes to three bytes. Essentially, this is a layout optimization.

I'll make a note to fix that comment to represent this better.

Reply:
> However, moving the Option causes the enum to grow in size, from two bytes to three bytes. Essentially, this is a layout optimization.

@pitaj Would you get the same layout optimization if you put `n_end_hashes` as the first field and then moved the option to that field? Nothing else in the code should care about the field order, so you can reorder the fields for layout optimization.

Reply:
Are you suggesting the following? 
```rust
GuardedStr { n_end_hashes: Option<u8>, n_start_hashes: NonZeroU8 }
```
The order doesn't make a difference. The discriminant can't be stored in a separate field.

Reply:
@pitaj :facepalm: Sorry, I managed to completely gloss over that one was `NonZeroU8` and the other was just `u8`. Please ignore.